### DISCUSSION TITLE
How To handle Initialization depending on two future providers

### DISCUSSION THREAD

**Discussion by penhorwood** [rrousselGit/riverpod#1722] [General]
I am super new to RiverPod so maybe I just missed how to do this.  In my app I am using Firebase Cloud Messaging, Isar for the database, Sentry and some encryption with keys stored in Secure Prefs.  All this stuff needs to be initialized.  I was using a singleton initialization class to handle this which does work.  What is the suggested way with RiverPod?  The initialization has a definite order as some parts depend on others.

**Comment by penhorwood** [comment]
This situation, which I found in issues, is what I am trying to accomplish.
https://github.com/rrousselGit/riverpod/issues/329#issuecomment-955237107

> I want:
> 
>     * to use riverpod providers as dependencies when initializing `prefs` and `auth`.
> 
>     * to call async code for preparing `prefs` and `auth`.
> 
>     * to override them in the container
>       as non-async providers (so that providers and other code that depend on them will not have to handle async complexities).
> 
> 
> Since a container only allows updating of overrides, I must initially override `prefs` and `auth` even though I do not have values for them, and later when I do I can update the container with the override values.

**Comment by penhorwood** [comment]
In Flutter most database initialization, store prefs, etc are using asynchronous methods therefore this seems like a normal use case.  What I can't find is how to go about it.

**Comment by penhorwood** [comment]
If you find yourself here I finally answered my own question via these blogs:

```
https://github.com/bizz84/flutter-tips-and-tricks/blob/main/tips/0056-async-init-provider-overrides/index.md
https://github.com/rrousselGit/riverpod/issues/295
https://codewithandrea.com/articles/riverpod-initialize-listener-app-startup/
```

**Comment by rohintoncollins** [comment]
@penhorwood  Some example code would be helpful here.
I have come across the same issue. I am in a widget and I need to wait for 2 async providers to complete. My first instinct is to create a third async provider which waits on the 1st 2. Then the widget can just wait on this 3rd provider.

**Reply by penhorwood** [reply]
I did basically what you suggested.  I created an Initialization class that handles every Initialization for my app.  Then I created methods for each Initialization.  Then I created an Initialization provider that makes sure it all happens.

I have one Initialization method I named primer that is ran by the provider.  The primer method makes sure everything gets ran in the correct order.

**Comment by gdurandrexel** [comment]
I did not try it, but maybe you could use utils like this (based on flutter\...\lib\async\future_extensions.dart).

```dart
import 'package:hooks_riverpod/hooks_riverpod.dart';

extension AsyncValueRecord2<T1, T2> on (AsyncValue<T1>, AsyncValue<T2>) {
  AsyncValue<(T1, T2)> get watch {
    if ($1.hasError) {
      return AsyncError($1.error!, $1.stackTrace!);
    }
    if ($2.hasError) {
      return AsyncError($2.error!, $2.stackTrace!);
    }

    if ($1.isLoading || $2.isLoading) {
      return const AsyncLoading();
    }

    return AsyncData(($1.requireValue, $2.requireValue));
  }
}

extension AsyncValueRecord3<T1, T2, T3> on (
  AsyncValue<T1>,
  AsyncValue<T2>,
  AsyncValue<T3>
) {
  AsyncValue<(T1, T2, T3)> get watch {
    if ($1.hasError) {
      return AsyncError($1.error!, $1.stackTrace!);
    }
    if ($2.hasError) {
      return AsyncError($2.error!, $2.stackTrace!);
    }
    if ($3.hasError) {
      return AsyncError($3.error!, $3.stackTrace!);
    }

    if ($1.isLoading || $2.isLoading || $3.isLoading) {
      return const AsyncLoading();
    }

    return AsyncData(($1.requireValue, $2.requireValue, $3.requireValue));
  }
}
```

and you can use it like so:
```dart
Widget build(BuildContext context, WidgetRef ref) {
    AsyncValue<First> firstAsync = ref.watch(firstProvider);
    AsyncValue<Second> secondAsync = ref.watch(secondProvider);

    AsyncValue<(First, Second)> combined = (firstAsync, secondAsync).watch;
...
}
```
